numba/unicode.py

```diff
@@ -531,6 +553,36 @@ def find_impl(a, b):
         return find_impl
 
 
+@overload_method(types.UnicodeType, 'count')
+def unicode_count(src, sub, start=None, end=None):
+
+    if not (start is None or isinstance(start, (types.Omitted, types.Integer, types.Optional))):
```

Comment:
Unit tests will need writing to ensure that the type rejection logic is correct.

Reply:
Sorry for my confusion but I am bit confused here. So are you suggesting that we should have a separate function for typechecking and then separate unit test for testing that funciton?

Reply:
@stuartarchibald we have several other function in unicode.py that do typechecking in similar ways. So in my opinion if we create a separate function then it should be flexible enough so that it can be used everywhere or we can keep it this way to maintain code format. what do you think?

Reply:
I think that if it is going to be used repeatedly, pulling the function out to module level makes sense.

Reply:
> Sorry for my confusion but I am bit confused here. So are you suggesting that we should have a separate function for typechecking and then separate unit test for testing that function?

I do not mind how the type checking is implemented, but it seems like a separate function would be useful as exactly the same checks are needed for `start` and `end`. There should be separate unit tests to check that compilation fails gracefully during type inference for unsupported/invalid types, an example is here:
https://github.com/numba/numba/blob/37ab1420fa87d02b635991b32f0faf5c65e8fc4d/numba/tests/test_unicode.py#L465-L470
this checks the multiplication of a string by a floating point value fails during typing.

Hope this helps.

Reply:
Oh okay! Got it. I'll make it in a separate function then.

Reply:
@stuartarchibald I have a doubt about this test showed as an example.  What does the following line do?
 self.assertIn('Invalid use of Function(<built-in function mul>)', str(raises.exception))

Can you explain briefly? Thanks!

Reply:
```python
self.assertIn('Invalid use of Function()', str(raises.exception))
```
this line is calling the `assertIn` function which asserts something contains something else, `str(raises.exception)` is converting the exception captured in the `raises` object yielded from the `with self.assertRaises` into a string so that it can be compared to an expected string. This is a convenient way of checking that the right class of exception was raised, along with the contents of the exception message being correct.